refactor: clean up `LedgerEntry.cpp`

mvadari opened the pull request:
## High Level Overview of Change

This PR heavily restructures `LedgerEntry.cpp` to make it easier to read and understand. 

Instead of having a nested series of if-elses, which is very difficult to parse, each object now has 3 elements in an array: the parameter name, a parsing function, and the expected ledger entry type.
```
{jss::account_root, parseAccountRoot, ltACCOUNT_ROOT},
```

This allows the parsing logic to be greatly simplified by returning early in each of the parsing functions, instead of needing to fall through with a lot of else statements - and this also avoids a lot of extra indenting (which in turn makes the code easier to read). It also makes it easier to audit what ledger entries are missing (e.g. NFT Offers are missing right now), and avoids bugs from forgetting to specify the expected ledger entry type.

It does not change any of the individual ledger entry parsing code, to make this PR easier to follow and review.

### Context of Change

`LedgerEntry.cpp` was a very messy file. I caught a bug in a PR that would have been avoided by this sort of structure.

### Type of Change

- [x] Refactor (non-breaking change that only restructures code)

### API Impact

No change in functionality.

## Test Plan

Tests remained mostly unchanged. We may want some basic performance regression tests, but the performance effect should be pretty minimal.

## Future Tasks

* Add parsing helper functions so that each new ledger entry parser doesn't need to reinvent the wheel.

mvadari commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
+        bool found = false;
+        for (const auto& ledgerEntry : ledgerEntryParsers)
         {
-            if (context.params.isMember("params") &&
```
This was only used for the CLI, so the CLI logic was moved to `RPCCall` instead, where it's cleaner and easier to read.

thejohnfreeman commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
+    }
+}
+
+using FunctionType =
+    std::function<std::optional<uint256>(Json::Value const&, Json::Value&)>;
```
I don't think you need runtime polymorphism here.
```suggestion
using FunctionType =
    std::optional<uint256>(*)(Json::Value const&, Json::Value&);
```

mvadari replied:
I'm unfamiliar with the pros and cons of `std::function` vs C-style pointers, I just assumed the norm was the more modern option.

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
+        return uNodeIndex;
+    }
+    // clang-format off
+    if (
```
This can be improved by splitting to some constants and/or helper function(s) 👍 Can be done in a later PR though.

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
+    // four strings defining the bridge (locking_chain_door,
+    // locking_chain_issue, issuing_chain_door, issuing_chain_issue)
+    // and the claim id sequence number.
+    auto lockingChainDoor = parseBase58<AccountID>(
```
Could be `const`?

mvadari replied:
Fixed in https://github.com/XRPLF/rippled/pull/5199/commits/3d5ef0fb972cf519f9c890126e2364e5316a014e

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
-            }
+    if (valid && claim_id[jss::xchain_owned_claim_id].isIntegral())
+    {
+        auto seq = claim_id[jss::xchain_owned_claim_id].asUInt();
```
Can be `const`

mvadari replied:
Fixed in https://github.com/XRPLF/rippled/pull/5199/commits/3d5ef0fb972cf519f9c890126e2364e5316a014e

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
+    if (!lpLedger)
+        return jvResult;
+
+    static LedgerEntry ledgerEntryParsers[] = {
```
nit: Consider using `std::array` over C-arrays

mvadari replied:
Fixed in https://github.com/XRPLF/rippled/pull/5199/commits/3d5ef0fb972cf519f9c890126e2364e5316a014e - not sure I like it, though

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -837,43 +889,51 @@ doLedgerEntry(RPC::JsonContext& context)
             // For apiVersion 2 onwards, any parsing failures that throw
             // this
```
nit: Looks like a formatting issue, maybe due to `clang-format` and column width limit changes in the past

mvadari replied:
Fixed in https://github.com/XRPLF/rippled/pull/5199/commits/3d5ef0fb972cf519f9c890126e2364e5316a014e

godexsoft reviewed (approved):
Looks like a huge improvement @mvadari , well done! 🥇 
I left a few nits. In general any of them could be left for another PR - as you wish.
Other than the ones I mentioned, i'd like to see `const` applied everywhere it can be. I'd also prefer to see some more spacing lines between `if` statements and stuff like that. Big `if` conditions could be broken down to helper functions/lambdas and/or constants. Some of them are pretty hard to read as they are now but I understand that's old code, not new one.

mvadari commented:
> Other than the ones I mentioned, i'd like to see `const` applied everywhere it can be. I'd also prefer to see some more spacing lines between `if` statements and stuff like that. Big `if` conditions could be broken down to helper functions/lambdas and/or constants. Some of them are pretty hard to read as they are now but I understand that's old code, not new one.

Thanks for the review! I added more spacing and I think I added `const` everywhere where applicable.

Definitely plan on adding more helper functions (every individual parser was written sort of individually, and had to reinvent a lot of wheels, and I think helper functions would do a lot to unify a lot of that). But wanted to save that for a separate PR to make this PR easier to review (since that code is already being moved around pretty significantly).

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -813,37 +830,41 @@ doLedgerEntry(RPC::JsonContext& context)
-    static LedgerEntry ledgerEntryParsers[] = {
-        {jss::index, parseIndex, ltANY},
-        {jss::account_root, parseAccountRoot, ltACCOUNT_ROOT},
+    static std::array<LedgerEntry, 21> ledgerEntryParsers = {
```
You can totally avoid specifying the size here like so: https://godbolt.org/z/6jETseadE
The downside is that you need to wrap each entry with its type (LedgerEntry{...}).

mvadari replied:
Fixed in https://github.com/XRPLF/rippled/pull/5199/commits/c255d460b42ac22e26ad29f94bd80dea97847c49

godexsoft commented on src/xrpld/rpc/handlers/LedgerEntry.cpp:
```diff
@@ -830,42 +830,38 @@ doLedgerEntry(RPC::JsonContext& context)
-    static std::array<LedgerEntry, 21> ledgerEntryParsers = {
-        LedgerEntry{jss::index, parseIndex, ltANY},
-        LedgerEntry{jss::account_root, parseAccountRoot, ltACCOUNT_ROOT},
+    static auto ledgerEntryParsers = std::to_array<LedgerEntry>({
```
Awesome 👍

mvadari commented:
@ximinez @bthomee this is ready to merge
